spiderexpress/model.py

```diff
@@ -235,9 +172,190 @@ class SamplerStateStore(Base):
     )
 
 
+def _merge_list_of_dicts(
+    session: orm.Session, data: List[Dict], factory: Callable[[Dict], Base]
+) -> None:
+    """Merge a list of dictionaries into the database."""
+    global _DATABASE_BUSY_  # pylint: disable=W0603
+    if _DATABASE_BUSY_ is True:
+        log.warning("Database is busy, skipping insert.")
+    _DATABASE_BUSY_ = True
+
+    for item in data:
+        session.merge(factory(item))
+
+    _DATABASE_BUSY_ = False
+
+
+def insert_layer_dense_edge(session: orm.Session, edge_type: str, data: List[Dict]):
+    """Insert a dense edge into the database."""
+    layer_counts = {}
+
+    def _factory_(item):
+        log.debug(f"Inserting dense edge for {item}")
+        source = item.get("source")
+        target = item.get("target")
+        layer_id = item.get("dispatch_with")
+        if layer_id not in layer_counts:
+            layer_counts[layer_id] = 0
+        else:
+            layer_counts[layer_id] += 1
+        id = f"{layer_id}:{source}-{target}"
+
+        return LayerDenseEdges(
+            id=id,
+            source=source,
+            target=target,
+            edge_type=edge_type,
+            layer_id=layer_id,
+            data=item,
+        )
+
+    _merge_list_of_dicts(session, data, _factory_)
+
+    _layer_count_str_ = ", ".join(
+        (f"{layer}: {count}" for layer, count in layer_counts.items())
+    )
+    log.info(f"Inserted {_layer_count_str_} dense edges.")
+
+
+def insert_layer_dense_node(
+    session: orm.Session, layer_id: str, node_type: str, data: List[Dict]
+):
+    """Insert a dense node into the database."""
+
+    def _factory_(item):
+        name = item.get("name")
+        id = f"{layer_id}:{name}"
+        return LayerDenseNodes(
+            id=id, name=name, layer_id=layer_id, node_type=node_type, data=item
+        )
+
+    _merge_list_of_dicts(session, data, _factory_)
+
+    log.info(f"Inserted {len(data)} dense node in layer {layer_id}")
+
+
+def insert_layer_sparse_node(
+    session: orm.Session, layer_id: str, node_type: str, data: List[Dict]
+):
+    """Insert a sparse node into the database."""
+
+    def _factory_sparse_node_(item):
+        name = item.get("name")
+        id = f"{layer_id}:{name}"
+
+        return LayerSparseNodes(
+            id=id, name=name, layer_id=layer_id, node_type=node_type, data=item
+        )
+
+    _merge_list_of_dicts(session, data, _factory_sparse_node_)
+    log.info(f"Inserted {len(data)} sparse nodes in layer {layer_id}.")
+
+
+def insert_layer_sparse_edge(
```

Comment:
lots of code duplication here and also for dense/sparse_node. Maybe rather have one function and sparse/dense as an argument.